staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -540,13 +542,29 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 		klog.V(1).InfoS("[graceful-termination] shutdown event", "name", httpServerStoppedListeningCh.Name())
 	}()
 
+	// we don't accept new request as soon as both ShutdownDelayDuration has
+	// elapsed and preshutdown hooks have completed.
+	preShutdownHooksHasStoppedCh := s.lifecycleSignals.PreShutdownHooksStopped
 	go func() {
-		defer drainedCh.Signal()
-		defer klog.V(1).InfoS("[graceful-termination] shutdown event", "name", drainedCh.Name())
+		defer klog.V(1).InfoS("[graceful-termination] shutdown event", "name", notAcceptingNewRequestCh.Name())
+		defer notAcceptingNewRequestCh.Signal()
 
 		// wait for the delayed stopCh before closing the handler chain
 		<-delayedStopCh.Signaled()
 
+		// Additionally wait for preshutdown hooks to also be finished, as some of them need
+		// to send API calls to clean up after themselves (e.g. lease reconcilers removing
+		// itself from the active servers).
+		<-preShutdownHooksHasStoppedCh.Signaled()
```

Comment:
this is wrong in the diagram

> //  AfterShutdownDelayDuration (delayedStopCh)   PreShutdownHooksStopped (preShutdownHooksHasStoppedCh)

now `preShutdownHooksHasStoppedCh` depends on `delayedStopCh` to be closed

Reply:
pre-shutdown hooks start running as soon as TERM signal is received, and it runs in the same go routine as `Run`
https://github.com/kubernetes/kubernetes/blob/78a4ba6af8c24b7d0e0efabfdbfed72fe4fa00ec/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L569-L577

`delayedStopCh` is closed in a separate goroutine
https://github.com/kubernetes/kubernetes/blob/78a4ba6af8c24b7d0e0efabfdbfed72fe4fa00ec/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L481-L494

Reply:
The above would mean that preshutdownHooks signal is blockedon delayedStopCh, which is not the case - it is independent:
https://github.com/kubernetes/kubernetes/pull/110026/files#diff-68eb390e148644c38174960be4d103c0e377e71eb540aa8c395f7d2278363973L567-L577

I guess probably there are better ways to represent it - maybe we should revisit it somehow.

But conceptually I think we're on the same page.

Reply:
ok, IC now the miscommunication, not big deal though